crates/cheatcodes/src/evm/record_debug_step.rs

```diff
@@ -0,0 +1,118 @@
+use alloy_primitives::U256;
+
+use foundry_evm_traces::CallTraceArena;
+use revm::interpreter::{InstructionResult, OpCode};
+
+use foundry_evm_core::buffer::{get_buffer_accesses, BufferKind};
+use revm_inspectors::tracing::types::{CallTraceStep, RecordedMemory};
+use spec::Vm::DebugStep;
+
+// A depth first traverse to flatten the recorded steps.
+pub(crate) fn flatten_call_trace(
+    root: usize,
+    arena: &CallTraceArena,
+    node_start_idx: usize,
+) -> Vec<&CallTraceStep> {
+    let mut out = Vec::new();
+    let mut nodes = Vec::new(); // Use a Vec as a stack
+    nodes.push(root);
+
+    while let Some(node_idx) = nodes.pop() {
+        // Pop from the end of the stack
+        let node = &arena.nodes()[node_idx];
+        if node_idx >= node_start_idx {
+            for step in &node.trace.steps {
+                out.push(step);
+            }
+        }
+        // Push children onto the stack in reverse order so that the first child is
+        // processed first
+        for &child_idx in node.children.iter().rev() {
+            nodes.push(child_idx);
+        }
+    }
```

Comment:
I think this is not correct because we will process deeper children later. it should go over nodes in the order in which they were created

can't we just iterate over `arena.nodes()[node_start_idx..]` here?

Reply:
fixed in https://github.com/foundry-rs/foundry/pull/8571/commits/562691fe90e6b71ffab84590dc7d40920d4e3392

Reply:
sorry, I think this might need a few more changes. right now this just firstly processes all node's steps, then all steps of its children. however, the correct ordering would be to place some of the children's steps in-between node steps

eg if we do call, the called contract does sload, and then we do an mload, we'd want to end up with `[{"op": "call", "depth": 1}, {"op": "sload", "depth": 2}, {"op": "mload", "depth": 1}]`

right now the deeper sload would be at the end I believe, even though it was executed in-between two opcodes

ideally we should just perform a simple DFS over arena iterating through `node.ordering`, appending steps and going deeper when we occur calls

Reply:
ah you're right. 

Fixed in https://github.com/foundry-rs/foundry/pull/8571/commits/941a5f130c0dbd855221a16bdd3c6a26a439c5db